lucille-plugins/lucille-pinecone/src/main/java/com/kmwllc/lucille/pinecone/indexer/PineconeIndexer.java

```diff
@@ -90,24 +114,29 @@ private void uploadDocuments(List<Document> documents, String embeddingField, St
                           entry -> Value.newBuilder().setStringValue(entry.getValue().toString()).build())))
                   .build()))
           .collect(Collectors.toList());
-
       upsertDocuments(upsertVectors, namespace);
-    }
-
-    if (mode.equalsIgnoreCase("update")) {
+    } else if (mode.equalsIgnoreCase("update")) {
       updateDocuments(documents, embeddingField, namespace);
+    } else if (mode.equalsIgnoreCase("delete")) {
+      deleteDocuments(documents, namespace, deletionPrefix);
     }
   }
 
-  private void upsertDocuments(List<VectorWithUnsignedIndices> upsertVectors, String namespace) throws IndexerException {
-    // max upsertSize is 2MB or 1000 records, whichever is reached first, regardless of dimension
-    // larger dimensions will mean smaller batch size limit
-    if (upsertVectors.size() > 1000) {
-      throw new IndexerException("max upsert size of each batch is 1000, reduce the batch size indexer configuration.");
+  private void deleteDocuments(List<Document> documents, String namespace, String additionalPrefix) throws IndexerException {
+    List<String> idsToDelete = documents.stream()
+        .map(doc -> doc.getId() + additionalPrefix)
+        .collect(Collectors.toList());
+
+    try {
+      index.deleteByIds(idsToDelete, namespace);
```

Comment:
In the serverless Pinecone, there are 3 ways to delete documents-- id, id prefix, namespaces. Deleting namespaces is similar to Solr deleting a collection, so not supporting that. Deletion by id & id prefix have the same syntax. In the case where user wants to delete record id "100", this will not delete records of id "1004", "10005", etc.

Reply:
outdated